### ISSUE TITLE
Enable caching of client-side resources

### ISSUE THREAD

**Issue by nwalters512** [PrairieLearn/PrairieLearn#3625]
Currently, on each page request, all client-side scripts and styles (Bootstrap CSS + JS, MathJax, lodash, etc.) is not cached on the client and must be loaded in their entirety on each page reload. This hurts the user experience for folks who are located a long way from our origin servers (e.g., anywhere outside the US) or who have slow network connections, or both.

This gets worse as we move towards running each question panel in a separate iframe. Without caching, we must load each resource once per question panel. By enabling caching, we make it possible to reasonably render a page with many submissions.

We should update PrairieLearn to allow many client-side resources to be cached.

## Resource types

There are several categories of resources we could conceivably handle, each of which must be handled differently:

* resources served directly out of `node_modules`
* resources served from PrairieLearn source (think local scripts and stylesheets)
* element resources, both PL-level and course-level
* files from `clientFilesCourse`, `clientFilesQuestion`, etc.

### `node_modules` resources

This includes things like `socket.io-client`, `jquery`, `fontawesome`, etc. - think things that individual pages would declare in their page `<head>`.

For these, we should append a `?cache=[key]` query param to all resource URLs, where `[key]` is some stable identifier for a particular version of the dependency, likely the dependency version (or the hash of the version). If that query param is present on a request, we can respond with `Cache-Control: public, max-age=31536000` to cache the resource essentially indefinitely. If the query param is not included, we'll continue to behave as we do today (don't send any `Cache-Control` header). When we install a new version of a dependency, the version will be different, and thus the `cache` query param will be different as well, so clients will request the new version and cache it instead.

Note that the value of the `cache` query param doesn't actually mean anything - if a client requests `cache=1.2.3` but version `2.3.4` is present on the server, version `2.3.4` is what will be served.

### Resources served by PrairieLearn

Caching resources served from the PL source code is slightly tricker, as we don't have a formal version to use as a cache buster. There are a few things we could use instead:

* The git commit hash of the version of PrairieLearn that is running.
* The hash of the file contents. This might be preferable to the git commit hash in case we start deploying PL without a `.git` directory present, such as in a Docker image.

As above, the cache buster would be added with `?cache=...`.

### Element resources

Elements *declare* the resources they depend on, which makes it easy to generate appropriate URLs for them on the fly. We can use a combination of the "`node_modules`" and "Resources served by PrairieLearn" resources outlined above to handle these assets.

* `nodeModulesScripts` and `nodeModulesStyles` can be handled with the `node_modules` strategy.
* `coreScripts`, `coreStyles`, `elementScripts`, and `elementStyles` can be handled with the "Resources served by PrairieLearn" strategy.

### `clientFilesCourse`/`clientFilesQuestion`/etc.

We can't reliably use a query param here, as these assets are generally accessed with a URL prefix that elements/questions/etc. append paths to. Instead, we can build the cachebuster into that URL prefix itself. Instead of a URL prefix like `/pl/course_instance/1/instructor/question/75/clientFilesQuestion/`, we can instead generate something like `/pl/course_instance/1/instructor/question/75/clientFilesQuestion/[course_git_hash]`, where `[course_git_hash]` is the Git hash of the version of the course that the resource was requested from. As with the `?cache=` strategy above, the hash wouldn't actually change which resource is served - it would simply be a way for us to force a browser to re-fetch a resource after a course is synced.

### Dev mode

During local dev, files can change on disk quite often. To avoid strange issues, we should disable caching in dev mode entirely. As all resources should be local and thus served very quickly, this should be acceptable.

### Dynamically-generated files

I *believe* dynamically-generated files should (must?) be deterministic, so it should be safe to allow clients to cache those as well. @mwest1066 would appreciate any input you have here.

**Comment by mwest1066** [comment]
Rather than using explicit cache busting, can we use ETags? Express supports them: https://expressjs.com/en/4x/api.html#etag.options.table

I'm actually not sure why this isn't working already for MathJax. It sounds like it should be enabled by default.

**Comment by nwalters512** [comment]
> Rather than using explicit cache busting, can we use ETags? Express supports them: https://expressjs.com/en/4x/api.html#etag.options.table
> 
> I'm actually not sure why this isn't working already for MathJax. It sounds like it should be enabled by default.

Good question - it is indeed already working for MathJax! Here's what I see for a MathJax request on current production PL:

```
Cache-Control: max-age=60
ETag: "31254-1c5fc537f6900"
Last-Modified: Sat, 26 Oct 1985 08:15:00 GMT
```

The problem with using `ETag` or `Last-Modified` is that you generally have to set `max-age=0` to ensure resources are always up-to-date and you *still* incur a network round trip to check if you need to load a new version or if you get a 304. For things like international users where latency is likely to be the main concern, ETags alone won't result in a significantly better experience for all users. For iframed questions, that's (# resources necessary to render a panel) * (# submissions + 2) network requests. And in my opinion, setting `max-age` anywhere higher than 0 when using `ETag` is unacceptable means you'll inevitably run into situations of using a mix of stale and fresh resources which could be confusing or even break the user experience.

The benefit of building a cache-busting-type thing into the URL is that it allows you to set `max-age` as high as it can possibly go because resources are considered immutable, and it allows you to skip cache validation because a given resource cannot ever change. This avoids the network round trip and makes for incredibly long lived caches.

**Comment by mwest1066** [comment]
> The benefit of building a cache-busting-type thing into the URL is that it allows you to set `max-age` as high as it can possibly go because resources are considered immutable, and it allows you to skip cache validation because a given resource cannot ever change. This avoids the network round trip and makes for incredibly long lived caches.

Ok, I buy this.

In terms of implementation, it seems to me that it might be better to put the cache key in the URL, rather than as a query string. For example, we could request mathjax as:

```
/node_modules/mathjax/startup.js?cache=8a9f983c
OR
/node_modules/8a9f983c/mathjax/startup.js
```

The problem with the query string is that `startup.js` immediately loads a million other files from relative URLs (more JS, fonts, all kinds of junk). I think these will all skip the query string, so they won't come from the cache?

Under the second scheme, I think that all the things that `startup.js` loads will also be correctly cached?

(incidentally, we currently load mathjax from a special `/MathJax` route for historical reasons, but I think we should switch to just loading it from `node_modules` like everything else)

**Comment by mwest1066** [comment]
In terms of implementation, are you thinking that we'd have a function like `locals.node_modules_path()` so we could do something like this in our EJS templates:
```
<script type="text/javascript" id="MathJax-script" src="<%= node_modules_path("mathjax") %>/startup.js"></script>
```
Here `node_modules_path()` would be able to translate to `/node_modules/8a9f983c/mathjax`?

**Comment by nwalters512** [comment]
Closed by #3649 and #5006.
